uwblueprint/focus-on-nature
# feat: add camp registration endpoints

## Notion ticket link

[Ticket Name](https://www.notion.so/uwblueprintexecs/Tasks-38b49f83f6ce4b07ae188dabb37631e8?p=d436ed7850ba48f28d8b3fd09cd66015)

* Leaving formQuestions to another PR


## Implementation description
* Added endpoints for updating Camp, and creating/updating/deleting a CampSession
   * PATCH /camp/
   * POST/PATCH/DELETE /camp/:campId/session
* Went through registration designs to figure out which APIs should be called and when
* All CampSession creation/deletion should be done through the /session endpoints, rather than modifying /camp/ directly
* `campers` and `waitlist` fields are never modified by these endpoints, to avoid payment issues


## Steps to test
1. Verify automated tests: (`docker exec -it scv2_ts_backend /bin/bash -c "yarn test"`)
2. Try updating Camp by calling `localhost:5000/camp/:campId`
3. Try adding CampSession to Camp with `localhost:5000/camp/:campId/session/`
4. Try updating CampSession in Camp with `localhost:5000/camp/:campId/session/:campSessionId`
5. Try deleting CampSession in Camp with `localhost:5000/camp/:campId/session/:campSessionId`

Test Camp Body:
```
{
    "ageLower": 15,
    "ageUpper": 30,
    "name": "test camp",
    "description": "description",
    "location": "canada",
    "fee": 25
}
```
Test CampSession Body:
```
{
    "campSessions": [
        {
            "capacity": 20,
            "active": true,
            "startTime": "12:00",
            "endTime": "17:00",
            "dates": [
                "Sun Mar 13 2022 20:01:14 GMT-0600 (Mountain Daylight Time)"
            ]
        }
    ]
}
```
Test FormQuestion Body: 
```
{
"question": "Q1",
"type": "Multiselect",
"required": true,
"description": "question description"
}
```


## What should reviewers focus on?
* Verify that the endpoints catch whether each CampSession is contained in the associated `Camp` (i.e. you must pass in the id of the Camp containing the session you want to delete)
* Verify that no functionality is missing for the registration flow


## Checklist

TODO
- [] be able to support bulk campSession creation, sort dates, check for overlap
